File: doubly_linked_list/DoublyLinkedList.java
package doubly_linked_list;

import java.util.Iterator;

public class DoublyLinkedList<T extends Comparable<T>> implements Iterable<T> {
    private class Node<E extends Comparable<E>> {
        private E data;
        private Node<E> next;
        private Node<E> prev;

        private Node(E data) {
            this(data, null);
        }

        private Node(E data, Node<E> next) {
            this(data, next, null);
        }

        private Node(E data, Node<E> next, Node<E> prev) {
            this.data = data;
            this.next = next;
            this.prev = prev;
        }
    }

    private class LLIterator implements Iterator<T> {
        private Node<T> cursor;

        LLIterator() {
            cursor = head;
        }

        public boolean hasNext() {
            return cursor != null;
        }

        public T next() {
            if (hasNext()) {
                T temp = cursor.data;
                cursor = cursor.next;
                return temp;
            }

            return null;
        }
    }

    private Node<T> head;
    private Node<T> tail;
    private int size;

    DoublyLinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    DoublyLinkedList(T data) {
        insert(data);
    }

    public boolean isEmpty() {
        return (head == null);
    }

    public void insert(T data) {
        size++;

        if (isEmpty()) {
            head = new Node<T>(data);
            tail = head;
            return;
        }

        Node<T> currentNode = head;

        while (currentNode != null && currentNode.data.compareTo(data) < 0) {
            currentNode = currentNode.next;
        }

        if (currentNode == null) {
            tail.next = new Node<T>(data, null, tail);
            tail = tail.next;
        } else {
            Node<T> newNode = new Node<T>(data);
            if (currentNode == head) {
                head = newNode;
            } else {
                currentNode.prev.next = newNode;
            }

            newNode.next = currentNode;
            newNode.prev = currentNode.prev;
            currentNode.prev = newNode;
        }
    }

    public T accessAtHead() {
        if (isEmpty()) {
            return null;
        }

        return head.data;
    }

    public T accessAtTail() {
        if (isEmpty()) {
            return null;
        }

        return tail.data;
    }

    public T accessAtIndex(int index) {
        if (index >= size || index < 0) {
            throw new IllegalArgumentException("Index " + index + " is out of bounds.");
        }

        Node<T> currentNode = head;

        for (int i = 0; i < index; i++) {
            currentNode = currentNode.next;
        }

        return currentNode.data;
    }

    public T accesByValue(T data) {
        Node<T> currentNode = head;

        while (currentNode != null && !currentNode.data.equals(data)) {
            currentNode = currentNode.next;
        }

        if (currentNode == null) {
            return null;
        }

        return currentNode.data;
    }

    public T removeAtHead() {
        if (isEmpty()) {
            return null;
        }

        size--;

        T data = head.data;

        if (head == tail) {
            tail = null;
        }

        head = head.next;
        head.prev = null;

        return data;
    }

    public T removeAtTail() {
        if (isEmpty()) {
            return null;
        }

        size--;

        T data = tail.data;

        if (head == tail) {
            head = null;
        }

        tail = tail.prev;
        tail.next = null;

        return data;
    }

    public T removeAtIndex(int index) {
        if (index >= size || index < 0) {
            throw new IllegalArgumentException("Index " + index + " is out of bounds.");
        }

        Node<T> currentNode = head;

        for (int i = 0; i < index; i++) {
            currentNode = currentNode.next;
        }

        T data = currentNode.data;

        if (currentNode.prev != null) {
            currentNode.prev.next = currentNode.next;
        }

        if (currentNode.next != null) {
            currentNode.next.prev = currentNode.prev;
        }
        size--;

        return data;
    }

    public T removeByValue(T data) {
        Node<T> currentNode = head;

        while (currentNode != null && !currentNode.data.equals(data)) {
            currentNode = currentNode.next;
        }

        if (currentNode == null) {
            return null;
        }

        if (currentNode.prev != null) {
            currentNode.prev.next = currentNode.next;
        }

        if (currentNode.next != null) {
            currentNode.next.prev = currentNode.prev;
        }

        size--;

        return data;
    }

    public boolean contains(T data) {
        return data.equals(accesByValue(data));
    }

    public void clear() {
        head = null;
        size = 0;
    }

    public int getSize() {
        return size;
    }

    public String toString() {
        StringBuilder result = new StringBuilder();

        Node<T> currentNode = head;

        while (currentNode != null) {
            result.append(currentNode.data);

            if (currentNode.next != null) {
                result.append(" <-> ");
            }

            currentNode = currentNode.next;
        }

        result.append("\n");

        return result.toString();
    }

    public Iterator<T> iterator() {
        return new LLIterator();
    }

    public static void main(String[] args) {
        DoublyLinkedList<Integer> intList = new DoublyLinkedList<>();

        for (int i = 0; i < 15; i++) {
            intList.insert((17 * (i + 1) + 3) % 12);
            System.out.println("Inserting " + (17 * (i + 1) + 3) % 12);
        }

        System.out.print("\nOriginal list:\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.removeAtHead();
        System.out.print("\nremoveAtHead():\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.removeAtTail();
        System.out.print("\nremoveAtTail():\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.removeByValue(7);
        System.out.print("\nremoveByValue(7):\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.removeAtIndex(4);
        System.out.print("\nremoveatIndex(4):\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        try {
            intList.removeAtIndex(70);
        } catch (Exception e) {
            System.out.println(e.toString());
        }
        System.out.print("\nremoveatIndex(70)):\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.removeByValue(100);
        System.out.print("\nremoveByValue(100):\n" + intList);
        System.out.println("Head: " + intList.accessAtHead());
        System.out.println("Tail: " + intList.accessAtTail());
        System.out.println("Index 4: " + intList.accessAtIndex(4));
        System.out.println("Contains 7: " + intList.contains(7));
        System.out.println("Contains 6: " + intList.contains(6));
        System.out.println("Size: " + intList.getSize());

        intList.clear();
        System.out.println("There should be no integer list output after this line.\n\n");

        System.out.println(intList);

        DoublyLinkedList<String> strList = new DoublyLinkedList<>();

        for (int i = 0; i < args.length; i++) {
            strList.insert(args[i]);
            System.out.println("Inserting " + args[i]);
        }

        System.out.print("\nOriginal list:\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        strList.removeAtHead();
        System.out.print("\nremoveAtHead():\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        strList.removeAtTail();
        System.out.print("\nremoveAtTail():\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        strList.removeByValue("should");
        System.out.print("\nremoveByValue(\"should\"):\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        strList.removeAtIndex(4);
        System.out.print("\nremoveAtIndex(4):\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        try {
            strList.removeAtIndex(70);
        } catch (Exception e) {
            System.out.println(e.toString());
        }
        System.out.print("\nremoveAtIndex(70):\n" + strList);
        System.out.println("Head: " + strList.accessAtHead());
        System.out.println("Tail: " + strList.accessAtTail());
        System.out.println("Index 4: " + strList.accessAtIndex(4));
        System.out.println("Contains \"should\": " + strList.contains("should"));
        System.out.println("Size: " + strList.getSize());

        strList.clear();
        System.out.println("There should be no string list output after this line.\n\n");
        System.out.println(strList);
    }
}

File: doubly_linked_list/doubly_linked_list.c
#include "doubly_linked_list.h"

// Node creation and deletion.
node_t *create_node(void *data)
{
    node_t *node = malloc(sizeof(node_t));

    if( !node )
    {
        return NULL;
    }

    node->data = data;
    node->next = node->prev = NULL;

    return node;
}

// List creation and deletion.
linked_list_t *create_linked_list(int (*compare)(void *, void *),
                                  void (*print)(void *, void *),
                                  void (*delete)(void *))
{
    linked_list_t *list = malloc(sizeof(linked_list_t));

    if( !list )
    {
        return NULL;
    }

    list->comparator = compare;
    list->printer = print;
    list->deleter = delete;
    list->size = 0;
    list->head = list->tail = NULL;

    return list;
}

void delete_linked_list(linked_list_t **list)
{
    if( !list || !*list )
    {
        return;
    }

    clear_list(*list);
    free(*list);
    *list = NULL;
}

// Insertion function
void insert(linked_list_t *list, void *data)
{
    if( !list || !data )
    {
        return;
    }

    node_t *new_node = create_node(data);

    if( !new_node )
    {
        return;
    }

    ++list->size;

    if( !list->head )
    {
        list->head = list->tail = new_node;
        return;
    }

    node_t *node = list->head;
    while( node && list->comparator(node->data, data) < 0 )
    {
        node = node->next;
    }

    if( !node )
    {
        new_node->prev = list->tail;
        list->tail->next = new_node;
        list->tail = new_node;
    }
    else
    {
        if( node == list->head )
        {
            list->head = new_node;
        }
        else
        {
            node->prev->next = new_node;
        }

        new_node->next = node;
        new_node->prev = node->prev;
        node->prev = new_node;
    }
}

// Access functions
void *access_at_head(linked_list_t const *const list)
{
    return (list->head) ? list->head->data : NULL;
}

void *access_at_tail(linked_list_t const *const list)
{
    return (list->tail) ? list->tail->data : NULL;
}

void *access_at_index(linked_list_t const *const list, size_t const index)
{
    if( !list || index >= list->size )
    {
        return NULL;
    }

    size_t back_index = list->size - index;
    node_t *node = (back_index > index) ? list->head : list->tail;

    if( back_index > index )
    {
        for( size_t i = 0; i < index; ++i )
        {
            node = node->next;
        }
    }
    else
    {
        for( size_t i = 0; i < back_index; ++i )
        {
            node = node->prev;
        }
    }

    return node->data;
}

void *access_by_value(linked_list_t const *const list, void *data)
{
    if( !list || !data )
    {
        return NULL;
    }

    node_t *node = list->head;
    int cmp_result;

    while( node && (cmp_result = list->comparator(node->data, data)) < 0 )
    {
        node = node->next;
    }

    if( !node || cmp_result != 0 )
    {
        return NULL;
    }

    return node->data;
}

// Removal functions
void *remove_at_head(linked_list_t *list)
{
    if( !list || !list->head )
    {
        return NULL;
    }

    node_t *node = list->head;
    void *data = node->data;

    if( list->head == list->tail )
    {
        list->head = list->tail = NULL;
    }
    else
    {
        list->head = list->head->next;
        list->head->prev = NULL;
    }

    free(node);
    --list->size;

    return data;
}

void *remove_at_tail(linked_list_t *list)
{
    if( !list || !list->tail )
    {
        return NULL;
    }

    node_t *node = list->tail;
    void *data = node->data;

    if( list->head == list->tail )
    {
        list->head = list->tail = NULL;
    }
    else
    {
        list->tail = list->tail->prev;
        list->tail->next = NULL;
    }

    free(node);
    --list->size;

    return data;
}

void *remove_at_index(linked_list_t *list, size_t const index)
{
    if( !list || index >= list->size )
    {
        return NULL;
    }

    if( index == 0 )
    {
        return remove_at_head(list);
    }
    else if( index == list->size - 1 )
    {
        return remove_at_tail(list);
    }

    size_t back_index = list->size - index;
    node_t *node = (back_index > index) ? list->head : list->tail;

    if( back_index > index )
    {
        for( size_t i = 0; i < index; ++i )
        {
            node = node->next;
        }
    }
    else
    {
        for( size_t i = 0; i < back_index; ++i )
        {
            node = node->prev;
        }
    }

    void *data = node->data;

    node->next->prev = node->prev;
    node->prev->next = node->next;

    free(node);
    --list->size;

    return data;
}

void *remove_by_value(linked_list_t *list, void *data)
{
    if( !list || !data )
    {
        return NULL;
    }

    node_t *node = list->head;
    int cmp_result;

    while( node && (cmp_result = list->comparator(node->data, data)) < 0 )
    {
        node = node->next;
    }

    if( !node || cmp_result != 0 )
    {
        return NULL;
    }

    if( node == list->head )
    {
        return remove_at_head(list);
    }
    else if( node == list->tail )
    {
        return remove_at_tail(list);
    }

    void *result = node->data;
    node->prev->next = node->next;
    node->next->prev = node->prev;

    free(node);
    --list->size;

    return result;
}

void clear_list(linked_list_t *list)
{
    if( !list )
    {
        return;
    }

    node_t *node = list->head;

    while( node )
    {
        node_t *next = node->next;

        list->deleter(node->data);
        free(node);
        node = next;
    }

    list->head = list->tail = NULL;
    list->size = 0;
}

// Utility functions
void print_list(linked_list_t const *const list, FILE *fp)
{
    if( !list || !list->head || !fp )
    {
        return;
    }

    node_t *node = list->head;

    while( true )
    {
        list->printer(node->data, fp);

        if( node->next )
        {
            fprintf(fp, " <-> ");
        }
        else
        {
            fprintf(fp, "\n");
            break;
        }

        node = node->next;
    }
}

bool contains(linked_list_t const *const list, void *data)
{
    return access_by_value(list, data);
}

size_t get_size(linked_list_t const *const list)
{
    return (list) ? list->size : 0;
}